File: TMB_objective_functions/GenopairNLL.cpp
// Genopair negative log-likelihood function for TMB

#include <TMB.hpp>

template<class Type>
Type objective_function<Type>::operator() ()
{
  // Declare data inputs
  DATA_MATRIX(gpprobs);
  DATA_IMATRIX(sampyrinds);
  DATA_INTEGER(npairs);
  DATA_INTEGER(k);
  DATA_IVECTOR(srvygaps);
  DATA_INTEGER(fyear);
  DATA_IVECTOR(srvyyrs);
  DATA_INTEGER(alpha);
  
  // Declare parameter input
  PARAMETER_VECTOR(pars);
  
  // Unpack parameters and request standard error for lambda
  Type rho = pars(0);
  Type phi = pars(1);
  Type Nfinal = pars(2);
  Type lambda = rho + phi;
  ADREPORT(lambda);
  
  // Find superpopulation size to compare with POPAN estimates
  
  // Find lambda and phi with respect to gaps between surveys
  vector<Type> lambdagaps(k - 1);
  vector<Type> phigaps(k - 1);
  for(int i = 0; i < k - 1; i++) {
    lambdagaps(i) = pow(lambda, srvygaps(i));
    phigaps(i) = pow(phi, srvygaps(i));
  }
  
  // Find entry proportions
  vector<Type> pentvec(k);
  pentvec(0) = Type(1.0);
  Type cumlambda = Type(1.0);
  for(int t = 1; t < k; t++) {
    pentvec(t) = (lambdagaps(t - 1) - phigaps(t - 1)) * cumlambda;
    cumlambda *= lambdagaps(t - 1);
  }
  Type sumpentvec = pentvec.sum();
  for(int i = 0; i < k; i++) {
    pentvec(i) = pentvec(i) / sumpentvec;
  }
  
  // Find Ns as first N_t divided by first entry proportion and specify it as a
  // derived parameter to calculate and report the standard error
  Type Ns = (Nfinal / cumlambda) / pentvec(0);
  ADREPORT(Ns);
  
  // Set negative log likelihood to zero
  Type nll = Type(0.0);
  
  // Parent-offspring pair probabilities within samples
  
  // Declare variables for loop
  Type expNsurvyr;
  matrix<Type> prbPOPsmat(k, k);
  prbPOPsmat.setZero();

  // Loop over surveys
  for(int srvyind = 0; srvyind < k; srvyind++) {
    // Find the expected number alive at the sample year
    expNsurvyr = Nfinal / pow(lambda, fyear - srvyyrs(srvyind));
    
    // Probability of POPs within sample
    prbPOPsmat(srvyind, srvyind) = Type(2.0) / (expNsurvyr - Type(1.0)) * rho * 
      (Type(1.0) + phi) / (lambda - pow(phi, 2));
  }
  
  // Self and parent-offspring pairs between samples

  // Declare variables for loop
  Type expNsurvyr1, expNsurvyr2, prbPOPsbrnbtn;
  matrix<Type> prbSPsmat(k, k);
  prbSPsmat.setZero();
  int srvyyr1, srvyyr2, srvygap;

  // Loop over all but last survey
  for(int srvyind1 = 0; srvyind1 < k - 1; srvyind1++) {
    // Find first survey year
    srvyyr1 = srvyyrs(srvyind1);

    // Loop over surveys with greater indices than first
    for(int srvyind2 = srvyind1 + 1; srvyind2 < k; srvyind2++) {
      // Find second survey year
      srvyyr2 = srvyyrs(srvyind2);

      // Find gap between surveys.  Remember not necessarily consecutive
      srvygap = srvyyr2 - srvyyr1;

      // Find the expected numbers alive in survey years
      expNsurvyr1 = Nfinal / pow(lambda, fyear - srvyyr1);
      expNsurvyr2 = Nfinal / pow(lambda, fyear - srvyyr2);

      // Probability of SPs between samples
      prbSPsmat(srvyind1, srvyind2) = pow(phi, srvygap) / expNsurvyr2;
      
      // New probability of POPs between samples
      if(srvyyr1 + alpha < srvyyr2) {
        prbPOPsbrnbtn = (srvyyr2 - (srvyyr1 + alpha)) * 
          pow(lambda / phi, srvyyr1 + alpha);
      } else {
        prbPOPsbrnbtn = Type(0.0);
      }
      prbPOPsmat(srvyind1, srvyind2) = prbPOPsmat(srvyind1, srvyind1) * 
        (expNsurvyr1 - 1) / expNsurvyr2 * pow(phi, srvygap) + 
        Type(2.0) / expNsurvyr1 * (Type(1.0) - phi / lambda) * 
        pow(phi / lambda, srvyyr2) *
        ((pow(lambda / phi, srvyyr1 + Type(1.0)) - 
        pow(lambda / phi, std::min(srvyyr1 + alpha, srvyyr2) + Type(1.0))) / 
        (Type(1.0) - lambda / phi) + prbPOPsbrnbtn);
    }
  }
  ADREPORT(prbPOPsmat);
  ADREPORT(prbSPsmat);
  
  int inds1, inds2;
  Type prbPOP, prbSP;
  
  // Loop over genopairs
  for(int gpind = 0; gpind < npairs; gpind++) {
    // Get sample-year indices and kinship probabilities
    inds1 = sampyrinds(gpind, 0);
    inds2 = sampyrinds(gpind, 1);
    prbPOP = prbPOPsmat(inds1, inds2);
    prbSP = prbSPsmat(inds1, inds2);
    
    // Add negative log likelihood from genopair probabilities given kinships
    // and kinship probabilities in terms of parameters.
    nll = nll - log(
      (Type(1.0) - prbPOP - prbSP) * gpprobs(gpind, 0) +
        prbPOP * gpprobs(gpind, 1) + prbSP * gpprobs(gpind, 2)
    );
  }
  
  // Return negative log likelihood
  return nll;
}  

File: Functions/GenopairNLL.R
# Find genopair model negative log likelihood
GenopairNLLR <- function(
    params, k, f.year, srvy.yrs, alpha, gpps, smp.yr.ind.prs
) {
  # Unpack parameters
  rho <- params[1]
  phi <- params[2]
  N.fin <- params[3]
  lambda <- rho + phi
  
  # Rho must be strictly greater than zero for the closed form series below
  # to hold, otherwise nlminb gives warnings
  if (rho == 0) return(Inf)
  
  # Lambda minus phi-squared
  lmb.m.ph.sq = lambda - phi^2
  # Probability not new-born (phi over lambda)
  p.o.l = phi / lambda
  # Reciprocal of probability not new-born (phi over lambda)
  l.o.p = lambda / phi
  
  # Find the expected number alive at the survey year
  exp.N.s.yrs <- N.fin / lambda^(f.year - srvy.yrs)

  # All pairs in survey years
  exp.ns.APs.wtn = choose(exp.N.s.yrs, 2)
  
  # Parent-offspring pairs within survey years
  exp.ns.POPs.wtn = exp.N.s.yrs * rho * (1 + phi) / lmb.m.ph.sq
  
  # Survey-pair indices
  s.pr.inds = combn(k, 2)
  s.inds.1 = s.pr.inds[1, ]
  s.inds.2 = s.pr.inds[2, ]
  exp.N.s.yrs.1 = exp.N.s.yrs[s.inds.1]
  exp.N.s.yrs.2 = exp.N.s.yrs[s.inds.2]
  s.yrs.1 = srvy.yrs[s.inds.1]
  s.yrs.2 = srvy.yrs[s.inds.2]
  s.gaps = s.yrs.2 - s.yrs.1
  p.t.s.gs = phi^s.gaps
  
  # All pairs between pairs of surveys
  exp.ns.APs.btn = exp.N.s.yrs.1 * exp.N.s.yrs.2
  
  # Self-pairs between pairs of survey years
  exp.ns.SPs.btn = p.t.s.gs * exp.N.s.yrs.1
  
  # Parent-offspring pairs between survey years
  s.yrs.1.p.a = s.yrs.1 + alpha
  exp.ns.POPs.btn = 2 * exp.ns.POPs.wtn[s.inds.1] * p.t.s.gs +
    2 * exp.N.s.yrs.2 * (1 - p.o.l) * p.o.l^s.yrs.2 *
    ((l.o.p^(s.yrs.1 + 1) - l.o.p^(pmin(s.yrs.1.p.a, s.yrs.2) + 1)) /
       (1 - l.o.p) +
       ifelse(
         s.yrs.1.p.a < s.yrs.2,
         (s.yrs.2 - s.yrs.1.p.a) * l.o.p^s.yrs.1.p.a,
         0
       ))

  # Create matrices for probabilities of POPs and SPs over survey-pairs
  prb.POPs.mat <- prb.SPs.mat <- matrix(0, k, k)
  diag(prb.POPs.mat) = exp.ns.POPs.wtn / exp.ns.APs.wtn
  prb.POPs.mat[upper.tri(prb.POPs.mat)] = exp.ns.POPs.btn / exp.ns.APs.btn
  prb.SPs.mat[upper.tri(prb.SPs.mat)] = exp.ns.SPs.btn / exp.ns.APs.btn
  prb.UPs.mat = 1 - prb.POPs.mat - prb.SPs.mat
  
  # Find negative log-likelihood from genopair probabilities given kinships,
  # n_pairs x n_kinships (UP, POP, SP), survey-year index-pairs, n_pairs x 2,
  # and kinship probabilities given parameters, k x k for each kinship
  nll = -sum(log(
    gpps[, 1] * prb.UPs.mat[smp.yr.ind.prs] +
      gpps[, 2] * prb.POPs.mat[smp.yr.ind.prs] +
      gpps[, 3] * prb.SPs.mat[smp.yr.ind.prs]
  ))
  print(nll, digits = 20)
  
  # Return negative log likelihood
  nll
}